### PR TITLE
[REFACTORING] - Services, Similar Products, Carousel

### PR DESCRIPTION
**Pull request by castan1c**

### CODE REVIEW

File: src/app/components/similar-products/similar-products.component.ts [hunk 1]
```diff
@@ -1,9 +1,39 @@
...
+import { Component, OnDestroy, Input } from '@angular/core';
+import { AutoUnsubscribe } from 'ngx-auto-unsubscribe';
+import { Subscription } from 'rxjs';
 
+import { ProductService } from 'src/app/shared/services/product.service';
+import { ProductFormat } from 'src/app/app.enum';
+
+import { IProductShortInfo } from 'src/app/interfaces/product-short-info.interface';
+import { IProductSimilarOptions } from 'src/app/interfaces/product-similar-options.interface';
+import { IProduct } from 'src/app/interfaces/product.interface';
```
Comment (AlexLarusso) [inline src/app/components/similar-products/similar-products.component.ts]: Move to index.js

File: src/app/components/similar-products/similar-products.component.ts [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+
+@AutoUnsubscribe()
 @Component({
   selector: 'app-similar-products',
   templateUrl: './similar-products.html',
 })
-export class SimilarProductsComponent {
-  public category = 'similar';
+export class SimilarProductsComponent implements OnDestroy {
+  @Input() product: IProduct;
```
Comment (AlexLarusso) [inline src/app/components/similar-products/similar-products.component.ts]: `private/public` ?

File: src/app/components/similar-products/similar-products.html [hunk 3]
```diff
@@ -1,2 +1,2 @@
 <h1 class="content__title">Similar products</h1>
-<app-product-carousel [category]="category"></app-product-carousel>
\ No newline at end of file
+<app-product-carousel [productData]="similarProductsArray"></app-product-carousel>
```
Comment (AlexLarusso) [inline src/app/components/similar-products/similar-products.html]: You don't need to name it like `similarProductsArray`, just `similarProducts`

File: src/app/pages/home/home.component.ts [hunk 4]
```diff
@@ -1,20 +1,38 @@
...
+export class HomeComponent implements AfterViewInit, OnDestroy {
   @ViewChildren(ScrollAnchorDirective)
   private pageAnchors: QueryList<ScrollAnchorDirective>;
 
-  constructor(private scrollService: ScrollService) {}
+  public productList: Array<IProductShortInfo>;
+  public getProductsSub: Subscription;
+
+  constructor(private scrollService: ScrollService,
+    private productService: ProductService) { }
```
Comment (AlexLarusso) [inline src/app/pages/home/home.component.ts]: Please align:

```
  constructor(
    private scrollService: ScrollService,
    private productService: ProductService
) { }
```

It is more readable.

File: src/app/pages/product-details-page/product-details-page.component.ts [hunk 5]
```diff
@@ -1,11 +1,33 @@
-import { Component } from '@angular/core';
-import { ProductResolver } from '../../shared/services/product.resolver';
+import { Component, OnInit, OnDestroy } from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
+import { AutoUnsubscribe } from 'ngx-auto-unsubscribe';
 
+import { IProduct } from 'src/app/interfaces/product.interface';
+import { ProductService } from 'src/app/shared/services/product.service';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.component.ts]: Please sort. 

Angular's entities together, rxjs entities together, app together:
For example:

```
import { Component, OnInit, OnDestroy } from '@angular/core';

import { AutoUnsubscribe } from 'ngx-auto-unsubscribe';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { IProduct } from 'src/app/interfaces/product.interface';
import { ProductService } from 'src/app/shared/services/product.service';
```

App imports should be in the end

File: src/app/pages/product-details-page/product-details-page.component.ts [hunk 6]
```diff
@@ -1,11 +1,33 @@
...
   templateUrl: './product-details-page.html',
-  providers: [ProductResolver]
 })
 
-export class ProductDetailsPageComponent {
+export class ProductDetailsPageComponent implements OnInit, OnDestroy{
+  public productSource$: Observable<IProduct> = new Observable<IProduct>();
+
+  constructor(private productService: ProductService,
+    private route: ActivatedRoute) { }
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.component.ts]: Align

File: src/app/pages/product-details-page/product-details-page.component.ts [hunk 7]
```diff
@@ -1,11 +1,33 @@
...
+
+  public ngOnInit(): void {
+    console.log('pdp');
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.component.ts]: Remove, or implement `loggerService`

File: src/app/pages/product-details-page/product-details-page.html [hunk 8]
```diff
@@ -1 +1,3 @@
-<app-product-details></app-product-details>
+<app-similar-products *ngIf="productSource$ | async  as product"
+  [product]="product">
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.html]: Align

File: src/app/shared/product-carousel/product-carousel.component.ts [hunk 9]
```diff
@@ -1,9 +1,6 @@
-import { Component, OnInit, AfterViewInit, AfterViewChecked, ViewChild, HostListener, ElementRef, OnDestroy, Input } from '@angular/core';
+import { Component, OnInit, AfterViewInit, AfterViewChecked,
+  ViewChild, HostListener,ElementRef, Input } from '@angular/core';
```
Comment (AlexLarusso) [inline src/app/shared/product-carousel/product-carousel.component.ts]:
```
import {
  Component,
  OnInit,
  AfterViewInit,
  AfterViewChecked,
  ViewChild,
  HostListener,
  ElementRef,
  Input 
} from '@angular/core';
```

File: src/app/shared/product-carousel/product-carousel.component.ts [hunk 10]
```diff
@@ -32,24 +29,22 @@ const BREAK_POINTS = {
...
-@AutoUnsubscribe()
 @Component({
   selector: 'app-product-carousel',
   templateUrl: './product-carousel.html',
   styleUrls: ['./product-carousel.scss']
 })
-export class ProductCarouselComponent implements OnInit, AfterViewInit, AfterViewChecked, OnDestroy {
-  @Input() category: string;
+export class ProductCarouselComponent implements OnInit, AfterViewInit, AfterViewChecked {
+  @Input() productData: Array<IProductShortInfo>;
```
Comment (AlexLarusso) [inline src/app/shared/product-carousel/product-carousel.component.ts]: `private` or `public` ?

File: src/app/shared/services/http.service.ts [hunk 11]
```diff
@@ -12,4 +15,18 @@ export class HttpService {
   public getData(): Observable<Array<IProduct>> {
     return this.http.get<Array<IProduct>>(this.URL);
   }
+
+  public getProductById(id): Observable<IProduct> {
```
Comment (AlexLarusso) [inline src/app/shared/services/http.service.ts]: `(id: number)`

File: src/app/shared/services/http.service.ts [hunk 12]
```diff
@@ -12,4 +15,18 @@ export class HttpService {
...
+    return this.http.get<Array<IProduct>>(this.URL)
+      .pipe(map(products => products.find(product =>
+        product.id == id
+    )));
+  }
+
+  public getProductsByCategory(category: string): Observable<Array<IProduct>> {
+    return this.http.get<Array<IProduct>>(this.URL)
+      .pipe(map(products => products.filter(product =>
+        product.category == category
```
Comment (AlexLarusso) [inline src/app/shared/services/http.service.ts]: Always use strict comparison `===`

File: src/app/shared/services/product.resolver.ts [hunk 13]
```diff
@@ -7,14 +7,13 @@ import { ActivatedRouteSnapshot } from '@angular/router';
...
-import { ActivatedRoute } from '@angular/router';
 
 @Injectable()
 export class ProductResolver implements Resolve<IProduct> {
   constructor(private productService: ProductService) { }
 
   resolve(route: ActivatedRouteSnapshot): Observable<IProduct> {
-    return this.productService.getProduct(Number(route.paramMap.get('id')))
+    return this.productService.getProductById(route.paramMap.get('id'))
       .pipe(take(1));
```
Comment (AlexLarusso) [inline src/app/shared/services/product.resolver.ts]: `take(1)` or `first()` ?

Please check the difference

File: src/app/shared/services/product.service.ts [hunk 14]
```diff
@@ -2,17 +2,66 @@ import { Injectable } from '@angular/core';
...
 })
 export class ProductService {
   constructor(private httpService: HttpService) { }
 
-  public getProduct(id): Observable<IProduct> {
-    return this.httpService.getData().pipe(map(products =>
-      products.find(product => product.id === id)));
+  public getProducts(format: string = ProductFormat.full):
+    Observable<Array<any>> {
+      console.log(format);
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: No `console.log`s

File: src/app/shared/services/product.service.ts [hunk 15]
```diff
@@ -2,17 +2,66 @@ import { Injectable } from '@angular/core';
...
+  public getSimilarProducts(similarOptions: IProductSimilarOptions, format: string):
+    Observable<Array<any>> {
+      return this.httpService.getProductsByCategory(similarOptions.category)
+      .pipe(map(data => data.filter(product =>
+        product.category === similarOptions.category &&
+        product.sex === similarOptions.sex &&
+        product.id !== similarOptions.id)),
+          map(data => data.map(product =>
+            this.formatProduct(product, format))
+          )
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: Please simplify and move logic to separate function

File: src/app/shared/product-carousel/product-carousel.component.ts (hunk 9 above)
Comment (castan1c) [reply src/app/shared/product-carousel/product-carousel.component.ts]: Is this ok? 
```import {
  Component, OnInit, AfterViewInit, AfterViewChecked,
  ViewChild, HostListener, ElementRef, Input
} from '@angular/core';```

File: src/app/shared/services/http.service.ts (hunk 11 above)
Comment (castan1c) [reply src/app/shared/services/http.service.ts]: we are using string ids now, so i changed it to string type

File: src/app/shared/services/product.resolver.ts (hunk 13 above)
Comment (castan1c) [reply src/app/shared/services/product.resolver.ts]: first can receive callback function and throws an error if no value matched callback. We don't have any ErrorHandler at this moment. So i guess take(1) in this case is ok?

**Pull request merged**